### PR TITLE
Use the new JSDoc parser to handle @param tags

### LINKED ISSUE #6810: JsDoc optional params with [] syntax breaks JSDoc comments in signature help
The last two parameters below don't show the description in the signature help for that parameter, the first two do. It appears to be the `[name]` syntax that breaks this.

```javascript
/**
 * @param {string} p1 - A string param
 * @param {string?} p2 - An optional param 
 * @param {string} [p3] - Another optional param
 * @param {string} [p4="test"] - An optional param with a default value
 */
function f1(p1, p2, p3, p4){}
```

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
Fixes #6810 

We used to have two separate JSDoc Parameter Tag parsers -- one in services and one in parser. This PR removes the old tag parser in services and defers to the one in parser (which is more featureful and tested).

However, the parser doesn't extract doc comments (it leaves them as inter-node trivia), so we still need a function that extracts out the doc comments from a given JSDoc tag. This is now a general-purpose function `getCommentFromJsDocTag` that can do this for any kind of JSDoc tag (not just param tags).

### CODE REVIEW

File: src/services/services.ts [hunk 1]
```diff
@@ -375,6 +375,68 @@ namespace ts {
...
+                    }
+                    start = i;
+                }
+            }
+
+            if (start !== -1 && start < rangeEnd) {
+                parts.push(text.substr(start, rangeEnd - start));
+            }
+
+            const result: SymbolDisplayPart = {
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Why don't you just add a return type annotation and return the result directly?

File: src/services/services.ts [hunk 2]
```diff
@@ -375,6 +375,68 @@ namespace ts {
...
+                    break;
+                }
+            }
+
+            let i: number;
+            for (i = start; i < rangeEnd; i++) {
+                const ch = text.charCodeAt(i);
+                if (isLineBreak(ch)) {
+                    // Line break, push any non-empty content so far
+                    if (start !== -1 && start !== i) {
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: When would `start` be -1?

File: src/services/services.ts [hunk 3]
```diff
@@ -375,6 +375,68 @@ namespace ts {
...
+                        parts.push(text.substr(start, i - start));
+                    }
+
+                    // Keep scanning until we hit either a a non-ws non-* char
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: There's a stutter in this comment :smile:

Comment (DanielRosenwasser) [inline src/services/services.ts]: You probably meant "a non-ws or a non-asterisk char"

File: src/services/services.ts [hunk 4]
```diff
@@ -375,6 +375,68 @@ namespace ts {
...
+                    while (i < rangeEnd) {
+                        i++;
+                        const nextCh = text.charCodeAt(i);
+                        if ((nextCh === CharacterCodes.asterisk) || (isWhiteSpace(nextCh) || isLineBreak(nextCh))) {
+                            // Eat *s and whitespace
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Just write `continue;` to make it explicit

File: src/services/services.ts [hunk 5]
```diff
@@ -375,6 +375,68 @@ namespace ts {
...
+                        }
+                        else if (nextCh === CharacterCodes.at) {
+                            // Abort, we're running into another malformed jsdoc tag
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Malformed is not necessarily true - you could just be in the next JSDoc tag, right?

File: src/services/services.ts [hunk 6]
```diff
@@ -391,9 +453,25 @@ namespace ts {
                     // If it is parameter - try and get the jsDoc comment with @param tag from function declaration's jsDoc comments
                     if (canUseParsedParamTagComments && declaration.kind === SyntaxKind.Parameter) {
                         ts.forEach(getJsDocCommentTextRange(declaration.parent, sourceFileOfDeclaration), jsDocCommentTextRange => {
-                            const cleanedParamJsDocComment = getCleanedParamJsDocComment(jsDocCommentTextRange.pos, jsDocCommentTextRange.end, sourceFileOfDeclaration);
-                            if (cleanedParamJsDocComment) {
-                                addRange(jsDocCommentParts, cleanedParamJsDocComment);
+                            // Incoming function strips off leading /* and trailing */, so add those back (-2, +4 net)
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: "Incoming function"? Do you mean `parseIsolatedJSDocComment`? If so, make that explicit.

File: src/services/services.ts (hunk 5 above)
Comment (DanielRosenwasser) [inline src/services/services.ts]: Ah, never mind. You figure out the start of the next tag later on.

File: src/services/services.ts [hunk 7]
```diff
@@ -391,9 +453,25 @@ namespace ts {
...
+                            const content = parseIsolatedJSDocComment(sourceFileOfDeclaration.text, jsDocCommentTextRange.pos - 2, jsDocCommentTextRange.end - jsDocCommentTextRange.pos + 4);
+                            if (content && content.jsDocComment.tags) {
+                                const tags = content.jsDocComment.tags;
+
+                                for (let i = 0; i < tags.length; i++) {
+                                    const tag = tags[i];
+                                    if (tag.kind === SyntaxKind.JSDocParameterTag) {
+                                        const tagParamName = (tag as JSDocParameterTag).preParameterName || (tag as JSDocParameterTag).postParameterName;
+                                        // -2 because we don't want to see the trailing */
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Where is the `-2` you're talking about?

File: src/services/services.ts [hunk 8]
```diff
@@ -391,9 +453,25 @@ namespace ts {
...
+                                        const nextTagStart = (i === tags.length - 1) ? jsDocCommentTextRange.end : tags[i + 1].pos;
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Why even bother with this if you immediately stop at an `@`?

File: src/services/services.ts (hunk 5 above)
Comment (DanielRosenwasser) [inline src/services/services.ts]: But so does that mean that you'll always cut JSDoc comments off at an `@`?

Comment (RyanCavanaugh) [inline src/services/services.ts]: Only if the @ occurs at the start of a line (which is basically indistinguishable from a malformed comment)

File: src/services/services.ts (hunk 8 above)
Comment (RyanCavanaugh) [inline src/services/services.ts]: The @ might not be the range end

**Comment by billti** [comment]
Testing this with a JsDoc comment such as `p1` below it breaks the function tooltip (it shows the tooltip as `"@hotmail.com for more info"`) :

```javascript
/**
 * @param {string} p1 - A string param. 
 * Contact billti@hotmail.com for more info
 * @param {string?} p2 - An optional param 
 * @param {string} [p3] - Another optional param. me@hotmail.
 * @param {string} [p4="test"] - An optional param with a default value
 * @return {string} This is the result
 */
function f1(p1, p2, p3, p4){}
```

It seems to be caused by getCleanedJsDocComment, not specifically these changes, but in general looks like this code often treats `@` anywhere as special char, not just after the initial optional whitespace and asterisk. Worth fixing up.

File: src/services/services.ts [hunk 9]
```diff
@@ -375,6 +375,68 @@ namespace ts {
...
+            const text = sourceFile.text;
+
+            const parts: string[] = [];
+            let start = tag.end + 1;
+
+            // Eat any leading ignored characters
+            while (start < rangeEnd) {
+                const ch = text.charCodeAt(start);
+                if (ch === CharacterCodes.asterisk || isWhiteSpace(ch)) {
+                    start++;
```
Comment (billti) [inline src/services/services.ts]: Adding the below to the `if` expression gets rid of the leading `-` in @param comments nicely for me :smile: 

```
|| (ch === CharacterCodes.minus && tag.kind === SyntaxKind.JSDocParameterTag)
```

**Comment by RyanCavanaugh** [comment]
I've spent three straight days on this and it's still not handling all the use cases we're inventing during development. This isn't a P1 for the 1.8 Salsa release so I'm going to close this and we can re-address at a later date unless there are substantial objections.

**Pull request closed without merging**